### PR TITLE
cleaned up builds for Couch and Cassandra so 'cluster' 

### PR DESCRIPTION
**Pull request by the-voice**

### CODE REVIEW

File: mache-couchbase/src/main/java/com/excelian/mache/couchbase/CouchbaseCacheLoader.java [hunk 1]
```diff
@@ -29,11 +29,11 @@
 
     private Class<K> keyType;
     private Class<V> valueType;
-    private Cluster cluster;
+    static private Cluster cluster;
```
Comment (jbowkett) [inline mache-couchbase/src/main/java/com/excelian/mache/couchbase/CouchbaseCacheLoader.java]: `private static` is more usual

File: mache-couchbase/src/main/java/com/excelian/mache/couchbase/builder/CouchbaseProvisioner.java [hunk 2]
```diff
@@ -21,18 +22,17 @@
...
     private String adminPassword;
     private SchemaOptions schemaOptions;
 
-    private CouchbaseProvisioner(CouchbaseEnvironment couchbaseEnvironment, BucketSettings bucketSettings,
-                                 List<String> nodes, String adminUser, String adminPassword,
+    private CouchbaseProvisioner(Cluster cluster, BucketSettings bucketSettings,
+                                 String adminUser, String adminPassword,
                                  SchemaOptions schemaOptions) {
-        this.couchbaseEnvironment = couchbaseEnvironment;
+        this.cluster = cluster;
```
Comment (jbowkett) [inline mache-couchbase/src/main/java/com/excelian/mache/couchbase/builder/CouchbaseProvisioner.java:34]: Does this work with no clustering options?....Or does Couch always have a cluster, even if there's one node?

File: mache-couchbase/src/main/java/com/excelian/mache/couchbase/builder/CouchbaseProvisioner.java [hunk 3]
```diff
@@ -99,10 +110,26 @@ public CouchbaseProvisionerBuilder withSchemaOptions(SchemaOptions schemaOptions
...
+        }
+
+        @Override
+        public AdminSettings withBucketSettings(BucketSettings bucketSettings) {
+            this.bucketSettings=bucketSettings;
+            return this;
         }
+
+        public CouchbaseProvisioner build() {
+            if(cluster==null)
```
Comment (jbowkett) [inline mache-couchbase/src/main/java/com/excelian/mache/couchbase/builder/CouchbaseProvisioner.java]: Join lines please (as per checkstyle)

Comment (jbowkett) [inline mache-couchbase/src/main/java/com/excelian/mache/couchbase/builder/CouchbaseProvisioner.java]: Actually, ensuring a cluster should probably be enforced by the builder - is it? - if so, bin this check

File: mache-couchbase/src/test/java/com/excelian/mache/couchbase/CouchbaseCacheLoaderMockedTest.java [hunk 4]
```diff
@@ -112,8 +122,10 @@ private void thenClusterManagerAndBucketCreated(CouchbaseCluster mockedCluster)
 
     private void givenCacheLoaderWith(SchemaOptions schemaOptions) {
         BucketSettings bucket = DefaultBucketSettings.builder().name("test").build();
-        loader = new CouchbaseCacheLoader<>(String.class, Object.class, bucket, DefaultCouchbaseEnvironment.create(),
-                Collections.singletonList("localhost"), "Admin", "Pass", schemaOptions);
+
+        //TODO: need to disconnect() cluster
```
Comment (jbowkett) [inline mache-couchbase/src/test/java/com/excelian/mache/couchbase/CouchbaseCacheLoaderMockedTest.java]: Still outstanding?

**Comment by tellisnz** [comment]
I unfortunately suggest we revert this.

By making client's provide the Cluster, they're now handling connectivity to the back end which is what mache should be in charge off. Also, by making cluster static in the cache loader, i believe we break JVMs with more than one Mache in them.

**Comment by the-voice** [comment]
Cluster static in the cache loader is a mistake.

The problem this is attempting to solve is that the providers CouchbaseProvisioner and CassandraProvisioner were creating both the cluster connection and the 'session/bucket' where the cluster connection should be shared between caches and threads. So I unpicked the code so that CouchbaseCacheLoader creates a new connection against an existing cluster. 

I'll pick this conversation up in Slack....

**Pull request merged**